### PR TITLE
[Sema] NFC: Simplify type coercion of ClosureExprs

### PR DESCRIPTION
**Pull request by davezarzycki**
* If we have an updatable ClosureExpr, then both fromEI.isAutoClosure() and toEI.isAutoClosure() must be false because the type system ensures that normal closures and functions do not convert to auto-closures.
* If we have a throwing ClosureExpr, then fromEI.throws() and toEI.throws() must be true, otherwise an erroneous throwing to non-throwing closure/function conversions would be possible. From a different perspective, updating the 'throws' bit would make sense if overloading on 'throws' wasn't possible.

### CODE REVIEW

**Comment by davezarzycki** [comment]
@swift-ci please test

**Comment by davezarzycki** [comment]
Hi @lattner – Sorry for the stray click that turned into a GitHub review request. Carry on :-)

File: lib/Sema/CSApply.cpp [hunk 1]
```diff
@@ -6317,24 +6317,16 @@ Expr *ExprRewriter::coerceToType(Expr *expr, Type toType,
...
-      // throwing subexpressions. We also don't want to change the convention
-      // of the original closure.
+      // If we have a ClosureExpr, then we can safely propagate the 'no escape'
+      // bit to the closure without invalidating prior analysis.
       auto fromEI = fromFunc->getExtInfo(), toEI = toFunc->getExtInfo();
       if (toEI.isNoEscape() && !fromEI.isNoEscape()) {
-        swift::AnyFunctionType::ExtInfo newEI(fromEI.getRepresentation(),
-                                        toEI.isAutoClosure(),
-                                        toEI.isNoEscape() | fromEI.isNoEscape(),
-                                        toEI.throws() & fromEI.throws());
```
Comment (DougGregor) [inline lib/Sema/CSApply.cpp]: This is such a weird computation. We take the "autoclosure" bit from the "to" type, "noescape" from either type (even though it's only on the "to" type based on the "if" statement above) and "throws" from both types (which makes no sense).

**Review by DougGregor — APPROVED** [review]
Definitely a lot cleaner, thank you!

File: lib/Sema/CSApply.cpp (hunk 1 above)
Comment (davezarzycki) [reply lib/Sema/CSApply.cpp]: Indeed. Don't forget that this goofiness only happened when the no escape bit and only the no escape bit needed propagating. Thanks for the review!

**Comment by lattner** [comment]
nice patch @davezarzycki !

**Pull request merged**